lib/puppetserver/ca/revoke_action.rb

```diff
@@ -0,0 +1,149 @@
+require 'puppetserver/ca/utils'
+
+require 'optparse'
+require 'openssl'
+require 'net/https'
+require 'json'
+
+module Puppetserver
+  module Ca
+    class RevokeAction
+      SUMMARY = 'Revoke a given certificate'
+      BANNER = <<-BANNER
+Usage:
+  puppetserver ca revoke [--help|--version]
+  puppetserver ca revoke CERTNAME
+
+Description:
+Given a valid certname, instructs the CA to revoke it over
+HTTPS using the local agent's PKI
+
+Options:
+BANNER
+
+      def self.parser(parsed = {})
+        OptionParser.new do |o|
+          o.banner = BANNER
+          o.on('--config', 'Path to puppet.conf') do |conf|
+            parsed['config'] = conf
+          end
+        end
+      end
+
+      def initialize(logger)
+        @logger = logger
+      end
+
+      def parse(args)
+        results = {}
+        parser = self.class.parser(results)
+
+        _, certnames, invalid = Utils.parse_without_raising(parser, args)
+
+        exit_code = nil
+        unless invalid.empty?
+          Utils.print_invalid_args(@logger, invalid, parser.help)
+          exit_code = 1
+        end
+
+        case certnames.length
+        when 0
+          @logger.err 'A certname to revoke is required'
+          @logger.err parser.help
+          exit_code = 1
+        when 1
+          results[:certname] = certnames.first
+        else
+          @logger.err 'Only one certname is allowed at a time'
+          @logger.err 'Received certnames:'
+          certnames.each {|name| @logger.err "    #{name}" }
+          exit_code = 1
+        end
+
+        return results, exit_code
+      end
+
+      # take the parsed arg hash from #parse and perform this action
+      # print information to @logger and return a valid exit code
+      def run(args)
+        certname = args[:certname]
+        config = args[:config]
+
+        errors = validate_file_paths(config)
+        unless errors.empty?
+          @logger.err 'Error:'
+          errors.each {|error| @logger.err "    #{error}" }
+          exit 1
+        end
+
+        puppet = PuppetConfig.parse(config)
+
+        revoke(certname, puppet.settings)
+      end
+
+      def make_store(cacert, cert, crl)
+        store = OpenSSL::X509::Store.new
+        store.add_file(cacert)
+        store.add_file(cert)
+        store.add_file(crl)
+
+        store
+      end
+
+      def make_uri(host, port, certname)
+        URI('https://' + host + ':' + port +
+            '/puppet-ca/v1/certificate_status/' +
+            certname +
+            '?environment=production')
```

Comment:
The old Puppet code required every request to include an environment, but in cases like certificates (where environment doesn't make sense) it would ignore the environment param (though still, unfortunately require it). I should double check to see if Puppet Server made the environment optional when it re-implemented the CA endpoints (and we should probably make it optional in Puppet 6 if not)

Reply:
Based on the fact that this code https://github.com/puppetlabs/puppet/blob/master/lib/puppet/rest/routes.rb works, I think it's optional now.

Reply:
Double checked, the docs say it's still required, but I couldn't see it in the code and testing manually worked w/o supplying it. Removing.

Reply:
We should probably update the docs for Puppet 6 then.